File: packages/node/src/providers/microsoft/claims-mapper.ts
import { type AuthUser } from "../../core/auth-user.js";
import { type RawClaims } from "../../core/claims.js";
import { type ClaimMappingConfig, type NormalizedMicrosoftConfig } from "../../config/types.js";
import { AuthenticationError } from "../../errors/index.js";

/** Default source claims for each field, in priority order, for Microsoft tokens. */
const MICROSOFT_DEFAULTS: Required<{ [K in keyof ClaimMappingConfig]: string[] }> = {
  // `oid` is the immutable object id; `sub` is stable but pairwise per app.
  id: ["oid", "sub"],
  // v2 access tokens lack `email`; UPN/preferred_username cover most cases.
  email: ["email", "preferred_username", "upn", "unique_name"],
  name: ["name"],
  // App roles land in `roles`; some directories surface group ids in `groups`.
  roles: ["roles"],
  // v2 delegated scopes in `scp`; v1 sometimes uses `scope`.
  scopes: ["scp", "scope"],
  tenantId: ["tid"],
};

function toSourceList(value: string | string[] | undefined): string[] | undefined {
  if (value === undefined) return undefined;
  return Array.isArray(value) ? value : [value];
}

/** First claim in `sources` whose value is a non-empty string. */
function firstString(claims: RawClaims, sources: readonly string[]): string | undefined {
  for (const source of sources) {
    const value = claims[source];
    if (typeof value === "string" && value.length > 0) {
      return value;
    }
  }
  return undefined;
}

/**
 * Collect a string array from the first matching source. Accepts either a real
 * array of strings or a space-delimited string (how Entra encodes `scp`).
 */
function stringArray(claims: RawClaims, sources: readonly string[]): string[] {
  for (const source of sources) {
    const value = claims[source];
    if (Array.isArray(value)) {
      return value.filter((v): v is string => typeof v === "string" && v.length > 0);
    }
    if (typeof value === "string" && value.length > 0) {
      return value.split(" ").filter((v) => v.length > 0);
    }
  }
  return [];
}

/** Resolve the effective source list for a field: custom override else default. */
function sourcesFor(field: keyof ClaimMappingConfig, config: ClaimMappingConfig): string[] {
  return toSourceList(config[field]) ?? MICROSOFT_DEFAULTS[field];
}

/**
 * Map validated Microsoft claims into the normalized {@link AuthUser}, honoring
 * any config-driven {@link ClaimMappingConfig} overrides.
 *
 * The mapper is the only place that knows Microsoft claim names — the rest of the
 * framework consumes the provider-neutral `AuthUser`. Swapping defaults here (or
 * via config) is how a different directory's claim shape is accommodated without
 * touching middleware or core.
 */
export function mapClaimsToUser(
  claims: RawClaims,
  config: NormalizedMicrosoftConfig,
  providerName: string,
): AuthUser {
  const mapping = config.claims;

  const id = firstString(claims, sourcesFor("id", mapping));
  if (id === undefined) {
    // Without a stable subject we cannot safely identify the principal.
    throw new AuthenticationError("Token is missing a usable subject identifier", {
      tried: sourcesFor("id", mapping),
    });
  }

  const email = firstString(claims, sourcesFor("email", mapping));
  const name = firstString(claims, sourcesFor("name", mapping));
  const tenantId = firstString(claims, sourcesFor("tenantId", mapping));
  const roles = stringArray(claims, sourcesFor("roles", mapping));
  const scopes = stringArray(claims, sourcesFor("scopes", mapping));

  return {
    id,
    ...(email !== undefined ? { email } : {}),
    ...(name !== undefined ? { name } : {}),
    roles,
    scopes,
    ...(tenantId !== undefined ? { tenantId } : {}),
    provider: providerName,
    claims: Object.freeze({ ...claims }),
  };
}

File: packages/go/provider/microsoft/claims_mapper.go
package microsoft

import (
	"github.com/avi-pathak/easy-sso/packages/go/config"
	"github.com/avi-pathak/easy-sso/packages/go/core"
	"github.com/avi-pathak/easy-sso/packages/go/internal/oidc"
	"github.com/avi-pathak/easy-sso/packages/go/ssoerr"
)

// microsoftDefaults are the default source claims for each field, in priority
// order, for Microsoft tokens.
var microsoftDefaults = config.ClaimMappingConfig{
	// oid is the immutable object id; sub is stable but pairwise per app.
	ID: []string{"oid", "sub"},
	// v2 access tokens lack email; UPN/preferred_username cover most cases.
	Email: []string{"email", "preferred_username", "upn", "unique_name"},
	Name:  []string{"name"},
	// App roles land in roles.
	Roles: []string{"roles"},
	// v2 delegated scopes in scp; v1 sometimes uses scope.
	Scopes:   []string{"scp", "scope"},
	TenantID: []string{"tid"},
}

// MapClaimsToUser maps validated Microsoft claims into the normalized AuthUser,
// honoring any config-driven claim-mapping overrides. It is the only place that
// knows Microsoft claim names; the rest of the framework consumes AuthUser.
func MapClaimsToUser(claims core.Claims, cfg config.NormalizedMicrosoftConfig, providerName string) (*core.AuthUser, error) {
	m := cfg.Claims

	idSources := oidc.SourcesFor(m.ID, microsoftDefaults.ID)
	id, ok := oidc.FirstString(claims, idSources)
	if !ok {
		return nil, ssoerr.NewAuthenticationError(
			"Token is missing a usable subject identifier",
			map[string]any{"tried": idSources},
		)
	}

	email, _ := oidc.FirstString(claims, oidc.SourcesFor(m.Email, microsoftDefaults.Email))
	name, _ := oidc.FirstString(claims, oidc.SourcesFor(m.Name, microsoftDefaults.Name))
	tenantID, _ := oidc.FirstString(claims, oidc.SourcesFor(m.TenantID, microsoftDefaults.TenantID))
	roles := oidc.StringArray(claims, oidc.SourcesFor(m.Roles, microsoftDefaults.Roles))
	scopes := oidc.StringArray(claims, oidc.SourcesFor(m.Scopes, microsoftDefaults.Scopes))

	return &core.AuthUser{
		ID:       id,
		Email:    email,
		Name:     name,
		Roles:    roles,
		Scopes:   scopes,
		TenantID: tenantID,
		Provider: providerName,
		Claims:   claims,
	}, nil
}
